---
title: "differential_expression_analysis_salmon"
output: pdf_document
date: "2024-01-30"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r imports, include=FALSE}
#BiocManager::install("edgeR")
library(edgeR)
library(tximport)
library(stringr)
library(tidyverse)
#BiocManager::install("GenomicFeatures")
library(GenomicFeatures)
```

specify files:
```{r}
get_name <- function(filename){
  new_name <- paste("align_reps_salmon_boostrap_A", filename, sep = "/")
  return(new_name)
}
```

specify files:
```{r}
create_filename <- function(filename){
 new_name <- str_split_1(filename, "_")[1]
  return(new_name)
}
```

```{r}
setwd("/Users/christina/Documents/Salmon")
sample_id <- lapply(list.files("align_reps_salmon_boostrap_A"), create_filename)
sample_id <- unlist(sample_id)
sample_id
``` 

```{r}
conditions <- c("control1", "UPF1_KD", "UPF1_rescue", "SMG6_KD", "SMG6_rescue")
conditions <- lapply(conditions, function(x) rep(x, 3))
conditions <- unlist(conditions)
conditions
```




```{r}
files <- lapply(list.files("/Users/christina/Documents/Salmon/align_reps_salmon_boostrap_A"), get_name)
files <- unlist(files)
names(files) <- sample_id
#files_a <- files[1:15]#batch A
#files_b <- files[16:length(files)]#batch B
```


## Load gene to transcript mapping from database
```{r}
#gtf <- "merged_mixed_transcriptome_gtf_gene_id_filtered.gtf"
txdb.filename <- "transcriptome_gtf.sqlite"
#txdb <- makeTxDbFromGFF(gtf)
#saveDb(txdb, txdb.filename)
```

```{r}
txdb <- loadDb(txdb.filename)
txdf <- select(txdb, keys(txdb, "GENEID"), "TXNAME", "GENEID")
tab <- table(txdf$GENEID)
txdf$ntx <- tab[match(txdf$GENEID, names(tab))]
```

## NMD transcript IDs: for comparison
```{r}
library(stringr)
con <- file("/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Input2023/NMD_transcripts_CDNA.fa")
data <- readLines(con = con)
close(con)
IDs <- str_extract(data,"(?>ENSG.+\\|)[[:alnum:]]+")
IDs <- IDs[!is.na(IDs)]
ID_list <- unlist(str_split(IDs, pattern = "\\|"))
NMD_transcript_IDs <- ID_list[c(FALSE, TRUE)] 
#write.table(data.frame(IDs),file="IDs.txt",row.names = FALSE, col.names=FALSE, quote = FALSE)
```

# Load data: CatchSalmon to obtain dispersion estimates, divide by dispersion estimates
```{r}
catchSalmon_counts <- catchSalmon(files)#get overdispersion estimates of batch A
y <- DGEList(counts=catchSalmon_counts$counts/catchSalmon_counts$annotation$Overdispersion, genes=catchSalmon_counts$annotation, group = conditions)
```
# Differential expression of transcripts in 2 batches

```{r}
keep <- filterByExpr(y)
y <- y[keep,,keep.lib.sizes=FALSE]
y <- normLibSizes(y)
y <- calcNormFactors(y)
design <- model.matrix(~y$samples$group)
colnames(design) <- levels(c)
y <- estimateDisp(y,design)
```

```{r}
plotBCV(y)
```

```{r}
plotMDS(y, label = y$samples$group)
```
## SMG6_KD vrsus control


```{r}
#quasilikelihood
#comapre 2 to 1: 1 being control, and 2 I believe to be SMG6_KD
fit <- glmQLFit(y,design)
qlf <- glmQLFTest(fit,coef=2)
topTags(qlf)
```
```{r}
qlf$table$FDR <- p.adjust(qlf$table$PValue, method="BH")
sum(qlf$table$FDR  < 0.05)
sign_transcripts <- qlf$table[qlf$table$FDR < 0.05, ]
sign_transcripts$TXNAME = rownames(sign_transcripts)
merged_df <- merge(sign_transcripts, txdf[,c("GENEID", "TXNAME")], by = "TXNAME")
sign_transcripts_up <- merged_df[merged_df$logFC > 0, ]
sign_transcripts_down <-  merged_df[merged_df$logFC < 0, ]

```

```{r}
sign_transcripts_up_1_2 <- merged_df[merged_df$logFC > 0.26303440583379, ]
length(rownames(sign_transcripts_up_1_2))
```


```{r}
length(rownames(sign_transcripts_up))
length(rownames(sign_transcripts_down))

sum(sign_transcripts_up$GENEID %in% sign_transcripts_down$GENEID)
length(unique(merged_df$GENEID))
```

```{r}
sum(sign_transcripts_up$TXNAME %in% NMD_transcript_IDs)
sum(sign_transcripts_down$TXNAME %in% NMD_transcript_IDs)
insign_tr <- qlf$table[qlf$table$FDR >= 0.05, ]
sum(rownames(insign_tr) %in% NMD_transcript_IDs)

```

```{r}
SMG6_KD_up <- sign_transcripts_up$TXNAME
```

```{r}
plotQLDisp(fit)
```


```{r}
plotMD(qlf)
```
```{r}
summary(decideTests(qlf))
```
```{r}
insign_tr <- qlf$table[qlf$table$FDR >= 0.05, ]
sum(rownames(insign_tr) %in% NMD_transcript_IDs)
```

##  SMG6 KD versus rescue

```{r}
#quasilikelihood
#comapre 3 to 2: and 2 I believe to be SMG6_KD and 3 the rescue
fit <- glmQLFit(y,design)
qlf <- glmQLFTest(fit,contrast= c(0, 1, -1, 0, 0))
topTags(qlf)
```

```{r}
plotMD(qlf)
```

```{r}
qlf$table$FDR <- p.adjust(qlf$table$PValue, method="BH")
sum(qlf$table$FDR  < 0.05)
sign_transcripts <- qlf$table[qlf$table$FDR < 0.05, ]
sign_transcripts$TXNAME = rownames(sign_transcripts)
merged_df <- merge(sign_transcripts, txdf[,c("GENEID", "TXNAME")], by = "TXNAME")
sign_transcripts_up <- merged_df[merged_df$logFC > 0, ]
sign_transcripts_down <-  merged_df[merged_df$logFC < 0, ]
```

```{r}
sum(sign_transcripts_up$GENEID %in% sign_transcripts_down$GENEID)
length(unique(merged_df$GENEID))
```

```{r}
sum(sign_transcripts_up$TXNAME %in% NMD_transcript_IDs)
sum(sign_transcripts_down$TXNAME %in% NMD_transcript_IDs)
insign_tr <- qlf$table[qlf$table$FDR >= 0.05, ]
sum(rownames(insign_tr) %in% NMD_transcript_IDs)
```

```{r}
summary(decideTests(qlf))
```

```{r}
sign_transcripts_up_1_2 <- merged_df[merged_df$logFC > 0.26303440583379, ]
length(rownames(sign_transcripts_up_1_2))
```

```{r}
plotBCV(y)
```
## UPF1 KD against control


```{r}
#quasilikelihood
#comapre 4 to 1: 1 being control, and 4 I believe to be UPF1_KD
fit <- glmQLFit(y,design)
qlf <- glmQLFTest(fit,coef=4)
topTags(qlf)
```


```{r}
plotMD(qlf)
```
```{r}
qlf$table$FDR <- p.adjust(qlf$table$PValue, method="BH")
sum(qlf$table$FDR  < 0.05)
sign_transcripts <- qlf$table[qlf$table$FDR < 0.05, ]
sign_transcripts$TXNAME = rownames(sign_transcripts)
merged_df <- merge(sign_transcripts, txdf[,c("GENEID", "TXNAME")], by = "TXNAME")
sign_transcripts_up <- merged_df[merged_df$logFC > 0, ]
sign_transcripts_down <-  merged_df[merged_df$logFC < 0, ]
```

```{r}
sum(sign_transcripts_up$GENEID %in% sign_transcripts_down$GENEID)
length(unique(merged_df$GENEID))
```

```{r}
sum(sign_transcripts_up$TXNAME %in% NMD_transcript_IDs)
sum(sign_transcripts_down$TXNAME %in% NMD_transcript_IDs)
insign_tr <- qlf$table[qlf$table$FDR >= 0.05, ]
sum(rownames(insign_tr) %in% NMD_transcript_IDs)
```

```{r}
UPF1_KD_up <- sign_transcripts_up$TXNAME
```

```{r}
summary(decideTests(qlf))
```


```{r}
sign_transcripts_up_1_2 <- merged_df[merged_df$logFC > 0.26303440583379, ]
length(rownames(sign_transcripts_up_1_2))
```


## UPF1KD against rescue
```{r}
fit <- glmQLFit(y,design)
qlf <- glmQLFTest(fit,contrast= c(0, 0, 0, 1, -1))
topTags(qlf)
```

```{r}
qlf$table$FDR <- p.adjust(qlf$table$PValue, method="BH")
sum(qlf$table$FDR  < 0.05)
sign_transcripts <- qlf$table[qlf$table$FDR < 0.05, ]
sign_transcripts$TXNAME = rownames(sign_transcripts)
merged_df <- merge(sign_transcripts, txdf[,c("GENEID", "TXNAME")], by = "TXNAME")
sign_transcripts_up <- merged_df[merged_df$logFC > 0, ]
sign_transcripts_down <-  merged_df[merged_df$logFC < 0, ]
```

```{r}
sum(sign_transcripts_up$GENEID %in% sign_transcripts_down$GENEID)
length(unique(merged_df$GENEID))
```

```{r}
plotMD(qlf)
```
```{r}
sum(sign_transcripts_up$TXNAME %in% NMD_transcript_IDs)
sum(sign_transcripts_down$TXNAME %in% NMD_transcript_IDs)
insign_tr <- qlf$table[qlf$table$FDR >= 0.05, ]
sum(rownames(insign_tr) %in% NMD_transcript_IDs)
```


```{r}
summary(decideTests(qlf))
```

```{r}
sign_transcripts_up_1_2 <- merged_df[merged_df$logFC > 0.26303440583379, ]
length(rownames(sign_transcripts_up_1_2))
```

## Check intersection of UPF1_KD and SMG6_KD

```{r}
length(SMG6_KD_up)
length(UPF1_KD_up)
intersection <- SMG6_KD_up[SMG6_KD_up %in% UPF1_KD_up]
length(intersection)
```

```{r}
length(intersection[intersection %in% NMD_transcript_IDs])
```

```{r}
counter = 0
for (txid in intersection){
  if (startsWith(txid, "MST")){
    counter = counter + 1
  }
}
print(counter)
```

```{r}
#y$genes[intersection, ]
```









# Perform analysis for batch B

specify files:
```{r}
create_filename <- function(filename){
 new_name <- str_split_1(filename, "_")[1]
  return(new_name)
}
```

```{r}
setwd("/Users/christina/Documents/Salmon")
sample_id <- lapply(list.files("align_reps_salmon_bootstrap_B"), create_filename)
sample_id <- unlist(sample_id)
sample_id
``` 



```{r}
get_name <- function(filename){
  new_name <- paste("align_reps_salmon_bootstrap_B", filename, sep = "/")
  return(new_name)
}
```


```{r}
files <- lapply(list.files("/Users/christina/Documents/Salmon/align_reps_salmon_bootstrap_B"), get_name)
files <- unlist(files)
names(files) <- sample_id
#files_a <- files[1:15]#batch A
#files_b <- files[16:length(files)]#batch B
```


```{r}
conditions <- c("control2", "SMG7_KD", "SMG7_rescue", "dKD", "dKD_SMG6_rescue", "dKD_SMG7_rescue")
conditions <- lapply(conditions, function(x) rep(x, 3))
conditions <- unlist(conditions)
```

# Load data: CatchSalmon to obtain dispersion estimates, divide by dispersion estimates
```{r}
catchSalmon_counts <- catchSalmon(files)#get overdispersion estimates of batch A
y <- DGEList(counts=catchSalmon_counts$counts/catchSalmon_counts$annotation$Overdispersion, genes=catchSalmon_counts$annotation, group = conditions)
```
# Differential expression of transcripts in 2 batches

```{r}
keep <- filterByExpr(y)
y <- y[keep,,keep.lib.sizes=FALSE]
y <- normLibSizes(y)
y <- calcNormFactors(y)
design <- model.matrix(~y$samples$group)
colnames(design) <- levels(c)
y <- estimateDisp(y,design)
```


```{r}
design
```






```{r}
plotMDS(y, label = y$samples$group)
```

```{r}
plotBCV(y)
```
## DKD agaisnt control

```{r}
#test control 2 against the double knockdown
fit <- glmQLFit(y,design)
qlf <- glmQLFTest(fit,coef=2)
topTags(qlf)
```


```{r}
plotQLDisp(fit)
```

```{r}
qlf$table$FDR <- p.adjust(qlf$table$PValue, method="BH")
sum(qlf$table$FDR  < 0.05)
sign_transcripts <- qlf$table[qlf$table$FDR < 0.05, ]
sign_transcripts$TXNAME = rownames(sign_transcripts)
merged_df <- merge(sign_transcripts, txdf[,c("GENEID", "TXNAME")], by = "TXNAME")
sign_transcripts_up <- merged_df[merged_df$logFC > 0, ]
sign_transcripts_down <-  merged_df[merged_df$logFC < 0, ]
```

```{r}
sum(sign_transcripts_up$GENEID %in% sign_transcripts_down$GENEID)
length(unique(merged_df$GENEID))
```

```{r}
plotMD(qlf)
```

```{r}
sum(sign_transcripts_up$TXNAME %in% NMD_transcript_IDs)
sum(sign_transcripts_down$TXNAME %in% NMD_transcript_IDs)
insign_tr <- qlf$table[qlf$table$FDR >= 0.05, ]
sum(rownames(insign_tr) %in% NMD_transcript_IDs)
```

```{r}
dKD_up <- sign_transcripts_up$TXNAME
```

```{r}
summary(decideTests(qlf))
```

```{r}
sign_transcripts_up_1_2 <- merged_df[merged_df$logFC > 0.26303440583379, ]
length(rownames(sign_transcripts_up_1_2))
```

# double knockdown against SMG6 rescue

```{r}
#test SMG6 rescue against double KD
fit <- glmQLFit(y,design)
qlf <- glmQLFTest(fit,contrast= c(0, 1, -1, 0, 0, 0))
topTags(qlf)
```

```{r}
qlf$table$FDR <- p.adjust(qlf$table$PValue, method="BH")
sum(qlf$table$FDR  < 0.05)
sign_transcripts <- qlf$table[qlf$table$FDR < 0.05, ]
sign_transcripts$TXNAME = rownames(sign_transcripts)
merged_df <- merge(sign_transcripts, txdf[,c("GENEID", "TXNAME")], by = "TXNAME")
sign_transcripts_up <- merged_df[merged_df$logFC > 0, ]
sign_transcripts_down <-  merged_df[merged_df$logFC < 0, ]
```

```{r}
sum(sign_transcripts_up$GENEID %in% sign_transcripts_down$GENEID)
length(unique(merged_df$GENEID))
```

```{r}
plotMD(qlf)
```

```{r}
sum(sign_transcripts_up$TXNAME %in% NMD_transcript_IDs)
sum(sign_transcripts_down$TXNAME %in% NMD_transcript_IDs)
insign_tr <- qlf$table[qlf$table$FDR >= 0.05, ]
sum(rownames(insign_tr) %in% NMD_transcript_IDs)
```

```{r}
dKD_SMG6_rescue_down <- sign_transcripts_down$TXNAME
```



```{r}
summary(decideTests(qlf))
```

```{r}
sign_transcripts_up_1_2 <- merged_df[merged_df$logFC > 0.26303440583379, ]
length(rownames(sign_transcripts_up_1_2))
```

#DKD against SMG7 rescue
```{r}
#test SMG7 rescue against doble KD
fit <- glmQLFit(y,design)
qlf <- glmQLFTest(fit,contrast= c(0, 1, 0, -1, 0, 0))
topTags(qlf)
```

```{r}
qlf$table$FDR <- p.adjust(qlf$table$PValue, method="BH")
sum(qlf$table$FDR  < 0.05)
sign_transcripts <- qlf$table[qlf$table$FDR < 0.05, ]
sign_transcripts$TXNAME = rownames(sign_transcripts)
merged_df <- merge(sign_transcripts, txdf[,c("GENEID", "TXNAME")], by = "TXNAME")
sign_transcripts_up <- merged_df[merged_df$logFC > 0, ]
sign_transcripts_down <-  merged_df[merged_df$logFC < 0, ]
```

```{r}
sum(sign_transcripts_up$GENEID %in% sign_transcripts_down$GENEID)
length(unique(merged_df$GENEID))
sum(sign_transcripts_up$TXNAME %in% NMD_transcript_IDs)
sum(sign_transcripts_down$TXNAME %in% NMD_transcript_IDs)
insign_tr <- qlf$table[qlf$table$FDR >= 0.05, ]
sum(rownames(insign_tr) %in% NMD_transcript_IDs)
```

```{r}
plotMD(qlf)
```

```{r}
summary(decideTests(qlf))
```

```{r}
sign_transcripts_up_1_2 <- merged_df[merged_df$logFC > 0.26303440583379, ]
length(rownames(sign_transcripts_up_1_2))
```


## SMG7 KD against control
```{r}
#test control 2 against the SMG7 KD
fit <- glmQLFit(y,design)
qlf <- glmQLFTest(fit,coef=5)
topTags(qlf)
```

```{r}
qlf$table$FDR <- p.adjust(qlf$table$PValue, method="BH")
sum(qlf$table$FDR  < 0.05)
sign_transcripts <- qlf$table[qlf$table$FDR < 0.05, ]
sign_transcripts$TXNAME = rownames(sign_transcripts)
merged_df <- merge(sign_transcripts, txdf[,c("GENEID", "TXNAME")], by = "TXNAME")
sign_transcripts_up <- merged_df[merged_df$logFC > 0, ]
sign_transcripts_down <-  merged_df[merged_df$logFC < 0, ]
```

```{r}
sum(sign_transcripts_up$GENEID %in% sign_transcripts_down$GENEID)
length(unique(merged_df$GENEID))
```

```{r}
plotMD(qlf)
```

```{r}
sum(sign_transcripts_up$TXNAME %in% NMD_transcript_IDs)
sum(sign_transcripts_down$TXNAME %in% NMD_transcript_IDs)
insign_tr <- qlf$table[qlf$table$FDR >= 0.05, ]
sum(rownames(insign_tr) %in% NMD_transcript_IDs)
```

```{r}
KD_SMG7 <- sign_transcripts_up$TXNAME
```



```{r}
summary(decideTests(qlf))
```

```{r}
sign_transcripts_up_1_2 <- merged_df[merged_df$logFC > 0.26303440583379, ]
length(rownames(sign_transcripts_up_1_2))
```



## SMG rescue against SMG7 KD
```{r}
#test SMG7 rescue against SMG7 KD
fit <- glmQLFit(y,design)
qlf <- glmQLFTest(fit,contrast= c(0, 0, 0, 0, 1, -1))
topTags(qlf)
```



```{r}
qlf$table$FDR <- p.adjust(qlf$table$PValue, method="BH")
sum(qlf$table$FDR  < 0.05)
sign_transcripts <- qlf$table[qlf$table$FDR < 0.05, ]
sign_transcripts$TXNAME = rownames(sign_transcripts)
merged_df <- merge(sign_transcripts, txdf[,c("GENEID", "TXNAME")], by = "TXNAME")
sign_transcripts_up <- merged_df[merged_df$logFC > 0, ]
sign_transcripts_down <-  merged_df[merged_df$logFC < 0, ]
```

```{r}
sum(sign_transcripts_up$GENEID %in% sign_transcripts_down$GENEID)
length(unique(merged_df$GENEID))
```

```{r}
plotMD(qlf)
```


```{r}
sum(sign_transcripts_up$TXNAME %in% NMD_transcript_IDs)
sum(sign_transcripts_down$TXNAME %in% NMD_transcript_IDs)
insign_tr <- qlf$table[qlf$table$FDR >= 0.05, ]
sum(rownames(insign_tr) %in% NMD_transcript_IDs)
```



```{r}
summary(decideTests(qlf))
```

```{r}
sign_transcripts_up_1_2 <- merged_df[merged_df$logFC > 0.26303440583379, ]
length(rownames(sign_transcripts_up_1_2))
```


# Intersection with batch B
```{r}
batch_b_intersect <- dKD_up[dKD_up %in% KD_SMG7]
length(batch_b_intersect)
sum(batch_b_intersect %in% NMD_transcript_IDs)
```


```{r}
total_intersect <- batch_b_intersect[batch_b_intersect %in% intersection]
length(total_intersect)
length(intersection)
```


```{r}
sum(total_intersect %in% NMD_transcript_IDs)
```

```{r}
union <- c(SMG6_KD_up, UPF1_KD_up, dKD_up, KD_SMG7)
length(union)
```

```{r}
length(intersection[intersection %in% NMD_transcript_IDs])
```

```{r}
counter = 0
for (txid in total_intersect){
  if (startsWith(txid, "MST")){
    counter = counter + 1
  }
}
print(counter)
```

```{r}
counter = 0
for (txid in union){
  if (startsWith(txid, "MST")){
    counter = counter + 1
  }
}
print(counter)
```

```{r}
length(union[union %in% NMD_transcript_IDs])
```
